Add logging infrastructure

axunonb opened the pull request:
Closes #838 

`Ical.Net`:
* ~~Added package `Microsoft.Extensions.Logging.Abstractions` to `Ical.Net`~~
* Introduced new logging classes and interfaces (`ILogger`, `ILoggerFactory`, `NullLogger`, `NullLoggerFactory` et al) in namespace `Ical.Net.Logging`
* Introduced `LoggingProvider` to support global logger factory configuration and create loggers
* `NullLoggerFactory` is the default factory

`Ical.Net.Tests.Logging`
* Added `TestLoggingManager`
* Added `MicrosoftLoggerAdapter` and `MicrosoftLoggerFactoryAdapter` to wrap `ILogger` and `ILoggerFactory` for `Microsoft.Extensions.Logging`

The `TestLoggingManager` supplies two type of factories:
* `Ical.Net.Logging.ILoggerFactory that can be used with as the factory for `Ical.Net.Logging.LoggingProvider`. It allows for handling log message from inside `Ical.Net`.
* `Microsoft.Extensions.Logging.ILoggerFactory` as a general usage logging factory in unit tests.

The log messages of the loggers created by the factories go to the same configurable output (memory, file, console).
Under the hood the `Serilog` nuget package is used.

**Usage:**
- Instantiate `TestLoggingManager` in unit or integration tests to capture and inspect log output generated by iCal.NET components and components from inside the unit tests.
- You can specify logging options (like minimum/maximum log levels and logger name patterns) to control which log messages are recorded.
- Retrieve log entries via the `Logs` property, which returns the most recent logs up to a configurable maximum count (`Options.MaxLogsCount`).
- Useful for verifying that expected log messages are produced during test execution, or for debugging test failures by examining log output.
- Logging can be disabled in case no debugger is attached (`Options.DebugModeOnly`) which is `true` by default.

---

**Example Usage:**
https://github.com/ical-org/ical.net/blob/d659711585e81d3f64ff1df08f06e64726379b48/Ical.Net.Tests/Logging.Tests/TestLoggingManagerTests.cs#L62-L99

axunonb committed:
Add logging infrastructure

`Ical.Net`:
* Added package `Microsoft.Extensions.Logging.Abstractions` to `Ical.Net`
* Introduced `LoggingProvider` to support global logger factory configuration.
* `NullLoggerFactory` is the default factory

`Ical.Net.Tests`:
* Added ``TestLoggingProvider`

**Target:**
- The `TestLoggingProvider` is designed for use in testing scenarios within the iCal.NET project.
- It enables logging operations by configuring log targets, which can be either in-memory (`MemoryTarget`) or file-based (`FileTarget`).
- The target (where logs are stored) is set during construction, allowing flexibility for different test requirements.

**Usage:**
- Instantiate `TestLoggingProvider` in unit or integration tests to capture and inspect log output generated by iCal.NET components.
- You can specify logging options (like minimum/maximum log levels and logger name patterns) to control which log messages are recorded.
- Retrieve log entries via the `Logs` property, which returns the most recent logs up to a configurable maximum count (`Options.MaxLogsCount`).
- Useful for verifying that expected log messages are produced during test execution, or for debugging test failures by examining log output.
- Logging can be disabled in case no debugger is attached (`Options.DebugModeOnly`)

---

**Example Usage:**
```csharp
// Create a provider with in-memory logging
var provider = new TestLoggingProvider();

// Use provider.LoggerFactory to create loggers in your tests
var logger = provider.LoggerFactory.CreateLogger("TestLogger");

// Log something
logger.LogInformation("Test message");

// Retrieve logs for assertions
var logs = provider.Logs.ToList();
```

minichma commented on Ical.Net.Tests/Logging/TestLoggingProviderBase.cs:
```diff
@@ -0,0 +1,142 @@
+        LoggerFactory = CreateFactory(config);
+
+        // Set the iCal.NET logging configuration
+        LoggingProvider.SetLoggerFactory(LoggerFactory);
```
Setting the global static logger factory as a side effect of instantiating a class introduces additional complexities. I.e. it prevents tests from running in parallel if multiple test cases instantiate the type. Maybe use something like ThreadLocal, i.e. some provider that is instantiated only once globally and that can be used to register individual listeners or sub-providers per thread.

axunonb replied:
Oh yes, right! `AsyncLocal<ILoggerFactory>` looks like the most up-to-date alternative.

minichma replied:
I recommend _not_ setting a global configuration (even if its AsyncLocal) in a constructor. I'd rather suggest to create the test logger in one step and set it as the current logger in a separate one. Something like this:

```cs
public void SomeTest() {
    var testLoggingFactory = new TestLoggingFactory();
    using var loggingReg = TestLoggingProvider.RegisterLoggingFactoryAsyncLocal(testLoggingFactory);
    ... test code
}
```

The TestLoggingProvider would be something like this:

```cs
public static class TestLoggingProvider {

    private class AsyncLocalLoggerFactory : ILoggerFactory {
        private readonly AsyncLocal<ILoggerFactory> _loggerFactories = new();
        ....
    }

    private static readonly AsyncLocalLoggerFactory _asyncLocalLoggerFactory = new();

    static TestLoggingProvider() {
         LoggingProvider.SetLoggerFactory(_asyncLocalLoggerFactory);
    }

    static IDisposable RegisterLoggingFactoryAsyncLocal(ILoggerFactory loggerFactory) {
        _loggerFactories.SetCurrent(loggerFactory);
        return <some IDisposable that unregisters the current logger factory>
    }
}
```

axunonb committed:
Refactor logging framework and update tests (#852)

## Ical.Net
- Removed dependency `Microsoft.Extensions.Logging.Abstraction`
- Introduced new logging classes and interfaces (`ILogger`, `ILoggerFactory`, `NullLogger`, `NullLoggerFactory` et al) in namespace `Ical.Net.Logging`
- Refactored `LoggingProvider` to utilize `AsyncLocal<ILoggerFactory?>` for better async context management.
- Added extension methods for simplified logging calls.

## Ical.Net.Tests
- Added `Microsoft.Extensions.Logging` package
- Enhanced `LoggingProviderUnitTests` with new tests for logger factory initialization and multi-threaded logging behavior.
- Modified `TestLoggingProviderBase` to correctly use `NullTarget`.
- Updated `LogLevel` mappings for compatibility with `Ical.Net.Logging` implementation.

axunonb committed:
Remove unused `ILoggerT`, implement `IDisposable` pattern for `NullLoggerFactory`

minichma commented on Ical.Net/Logging/LoggingProvider.cs:
```diff
@@ -0,0 +1,69 @@
+/// </summary>
+internal static class LoggingProvider // Make public when logging is used in library classes
+{
+    private static readonly AsyncLocal<ILoggerFactory?> _loggerFactory =
```
Whether `AsyncLocal` should be used is rather specific to the individual application. In Ical.Net we need it for testing, but in most applications the users might not want to configure it per async flow but rather globally. In `TimeZoneResolver` we have a similar situation and don't use `AsyncLocal`. I recommend to use the AsyncLocal only within the test project, i.e. set a specific ILoggerFactory there that wraps a `AsyncLocal` and lets individual test cases configure individual loggers per async flow.

In a future version it would be good to get rid of the static global altogether. Just like in TimeZoneResolver it would be good to introduce some context that holds configuration like the logger or the TZ provider. But that's not a topic for now.

minichma commented on Ical.Net.Tests/Logging/Adapters/MicrosoftLoggerFactoryAdapter.cs:
```diff
@@ -0,0 +1,67 @@
+    public void Dispose()
+    {
+        Dispose(true);
+        GC.SuppressFinalize(this);
```
Why would we suppress the finalizer?

minichma reviewed (commented):
Very nice! In order to keep the code lean, I would probably have gone for simple delegates rather than individual `ILoggerFactory` and `ILogger` types, but that's probably a matter of taste. Maybe consider moving the `AsyncLocal` to the test project to reduce overall complexity and stay in line with what we have in `TimeZoneResolver` (where we also have some global configuration).

axunonb committed:
Switch logging framework from NLog to Serilog (#853)

Updated project to remove NLog references and integrate Serilog.
Added flexible logging options in the configuration.

axunonb committed:
Remove obsolete TestLoggingManager CTOR overloads
